### DISCUSSION TITLE
20231213 Release Candidate State

### DISCUSSION THREAD

**Discussion by mythrz** [o3de/o3de#17248] [General]
Some notes regarding the release (https://github.com/o3de/o3de/commit/f0c150d75722b3302753972b28d73a8036b70b31) and developer (https://github.com/o3de/o3de/commit/0f4feac58aab88c40a48cc2d0e724722740f2c59) branches.
Most fixes were made in **release** instead of the developer branch, unsure why. So this post is related to the state of the release candidate only.

- Script Canvas is now correctly passing/exposing **most** of the properties from the canvas to the Editor-Component as well as the Lua Script (https://github.com/o3de/o3de/issues/16769#issuecomment-1824443423).

---

- Lua script is not passing/exposing array properties from the canvas to the Editor-Component. SCs has a similar issue with arrays already reported. Instead of opening a new issue, I added a comment here https://github.com/o3de/o3de/issues/17208#issuecomment-1853997322
![arrayListNotWorking](https://github.com/o3de/o3de/assets/44119535/522e7128-4e1a-4df3-adf5-4dec1e93326f)

---

- Interaction with shape-cast and Prefabs. Could not test it, because the previous point is the one responsible for it!

---

- There is still something going on with materials, shaders, and default models (They are now working https://github.com/o3de/o3de/issues/17144 but with warnings. e.g., sphere.fbx)... There are some cases (not all) where if you manually reprocess they fix the warnings. Notice the "Completed" dates in the image below (some are from today, others from November):
![Reprocessing](https://github.com/o3de/o3de/assets/44119535/af5cd65d-7385-4e6b-8532-fe2ef2293e28)
The asset processor is not reprocessing some assets automatically. Maybe provide a simple warning to ask the user if he/she wants to reprocess assets with errors and warnings again? "Asset Processor -> Settings" might go unnoticed.

---

- After manually re-reprocess. The remaining assets that have errors or warnings stayed more or less the same (https://github.com/o3de/o3de/issues/12617#issuecomment-1840967272 , https://github.com/o3de/o3de/issues/12984#issuecomment-1841019314 , https://github.com/o3de/o3de/issues/12616#issuecomment-1853694953 , https://github.com/o3de/o3de/issues/12611#issuecomment-1840986795). This is one of the small things that scare new users the most when they install O3DE for the first time. They expect a clean/green/completed jobs-list, not something with warnings and errors.

---

- Procedural generation with Lua is working (at least my custom script is https://www.youtube.com/watch?v=OuvQDdNUF6M&ab_channel=rzDmyth ). But there is a strange issue with some of my prefabs (was unsure of how I made them). Installed the PBR reference materials, vegetation and landscape gems but did not fix the issue. Both the .materials, .shareders and .materialtypes are all with errors. You cannot use the O3DE material Editor to open and edit the .materials and .materialtypes!
a) In Asset processor:
AssetUtils: Could not find asset for source file [C:\O3DE_StabilityTest\o3de\Gems\Atom\Tools\MaterialCanvas\Assets\MaterialCanvas\GraphData\Nodes\MaterialOutputs\BasePBR\HandleCastShadows.lua]
LuaMaterialFunctorSourceData: Could not load script file '@gemroot:MaterialCanvas@/Assets/MaterialCanvas/GraphData/Nodes/MaterialOutputs/BasePBR/HandleCastShadows.lua'
MaterialTypeAsset: Failed to create MaterialFunctor
b) In Game Console:
![ProcGene](https://github.com/o3de/o3de/assets/44119535/a11b5064-f881-4aaa-b18c-963cf5513378)

- Next day, I noticed the errors are coming from the Material Canvas (Preview). A few months back I tested and created multiple .material, .materialgraph and .shader (https://www.youtube.com/watch?v=6Zi7JMp_t1k&ab_channel=rzDmyth). At that time, I later used those on the procedural generation sample... Tried to create new ones with this release candidate, and only the Graph Node Config Document saved without an error. If you create blank files it also works (at least it did not give an error with a blank New-Material-Graph-Document).
![materialCanvasPreview](https://github.com/o3de/o3de/assets/44119535/6db0ba2e-b09f-448c-a50c-4e814568747e)

---

- Some of my custom-made Lua scripts have errors even when changes were not made to them. Most are associated with some kind of dependency error. e.g. of error, "13/12/2023 10:45 | ERROR | LuaBuilder: Did not find dependency  degrees/sec referenced by script. from AssetBuilder". Funny thing, the Lua script works fine, it runs as expected in-game... So I don't know if there is an issue on my machine in this case. Please try it on your machines, here is the code: https://github.com/mythrz/Assets/blob/main/Projects/ObjectsAndCameraRotations/Scripts/ObjectRotations.lua
![Screenshot 2023-12-13 112040](https://github.com/o3de/o3de/assets/44119535/3c4c9082-e962-49a4-b44c-75f7c957b2b3)

---

- Remote Tools Connection triggers Windows firewall. This gem allows Lua debugging with O3DE Lua Editor. Why is it asking for access to the network? Strange/suspicious behaviour for local development.

---

- Animation Editor is working with my current prototypes. There is something quite annoying (life quality improvement) that can lead to a considerable loss of time.
a) The Motion Set is always creating a new Default folder, which is always empty https://github.com/o3de/o3de/issues/10375#issuecomment-1852303343 ! If you do not delete it every time you open the Animation Editor, the Anim Graph will not start as expected. This is also a factor in human-made-mistakes since the professional might accidentally delete the wrong Motion Set... forcing the professional to manually add all motions again.
b) The CTRL+S and Save all are not saving as expected in certain scenarios: https://github.com/o3de/o3de/issues/13224
https://github.com/o3de/o3de/issues/14868 , which can, again, induce professionals in error.

---

I think I did not forget anything that I tested so far. I will now discuss some other notes/feedback unrelated to the testing of this branch, but might be interesting nevertheless:

- Blueprints (EU5) vs Script Canvas (O3DE). Script Canvas is difficult to maintain, share and document (e.g., https://github.com/o3de/o3de.org/issues/1550 ). I once tested https://blueprintue.com with Unreal Engine 5 and is a very good tool for addressing these limitations. I once asked Sébastien if it could be adapted to O3DE. He replied that he indeed installed and tested O3DE but was unable to have a copy-paste like Unreal Engine, so he stopped investigations after that. I am not a big fan of SC, but those who prefer it to Lua might consider this a good solution. o3de.org could then have SCs that could be copied and pasted instead of images that no one can read.

---

- Documentation is way behind... I would suggest taking a different approach to it. Maybe record a video when the feature is complete and added to the release branch; post it as "temporary" in a YouTube channel, and link it in the o3de.org. It does not need to be perfect or final, but it should be more organized. A simple sketch or place-holder would be enough.
There is one piece of documentation that I think should be urgently reviewed. **How to create gems and components** with and without dependencies on other gem/modules/external/internal libs; how to CMake and build only the gem instead of the entire engine; how to save the gem in a private and public repository; how to re-use that gem in a fresh O3DE (re)install. The engine best selling point is the Gems and Components, so this should always be a priority.
Also worth noting that C++23 with named modules is expected soonTM (https://www.youtube.com/watch?v=_x9K9_q2ZXE&ab_channel=CppCon&t=26m35s ), so it might be a good idea to have the basics/foundations well documented. This will allow professionals with a good understanding of Modules (but not O3DE) to help out.
There are other topics (e.g., lack of database good practices) but I think gems/components is the most important...

---

- Roadmap. I have read in Discord that there might be one next year. I guess it will be a GitHub-Project-Roadmap? I think this is good news for those of us (noobs) who are not in the loop of what is going on internally on each team. My suggestions:
a) Flag features that might break the normal functionality of the engine (e.g., DPE). I was caught off-guard when I synced the Dev-branch. If I had somewhere to check this, I would have installed it in another folder.
b) Basic notes like which branch is the feature being developed on; who is the main-dev responsible (in case he/she wants to be flagged with the name-tag on new bug-issues, and make this easier to merge similar/close issues); flags like "testing-phase" to warn phases like this one for release-point; if the feature is currently being worked on; state of the feature ; etc...
c) Unassigned tasks that enthusiasts could work on and are of interest for O3DE growth. E.g. of a Task: Improve Lua documentation of script events. https://github.com/o3de/o3de.org/pull/2505 ! Note that these tasks do not need to be overly complicated. Some could be just "grunt work".
d) Individual projects (solutions and gems). Each one of these should have at least a task/issue associated with it (I think having individual projects for each one is overkill). This would help members outside those projects to check if development is active or not; if there are incoming future improvements; which are the projects/gems that were reviewed by O3DE and are considered "safe"; easy to find gaps or missing content for someone who wants to start something new and no one else is working on. 
e.g., of 3rd-party projects that I have no idea of their current state, but are somewhat important/relevant. These are just a few: 
https://www.youtube.com/watch?v=xbhGbSbKpvM&ab_channel=Open3DEngine 
https://www.youtube.com/watch?v=QYZiP9p5LPM&list=PLCQwFpnHSZQjDIkT9EqCNDWepsh8sTGMm&ab_channel=Open3DEngine
https://www.youtube.com/watch?v=96js30gsouY&list=PLCQwFpnHSZQg_4lCmfXhCB5EjxieUbipk&index=45&ab_channel=Open3DEngine

---

- I think the "main-branch" on GitHub is meant to be stable and recommended for use, correct? If this is the case, I suggest only updating it when there are no issues with it (even when you release a new version, this "main" or hypothetical "very-stable" version is only updated after). I believe the O3DE reputation is in shambles because in every release there is something that goes unchecked and passes an insecure message to the outside community. I would like very much to have one GitHub branch that I can always trust to be battle-tested, I can easily recommend to others and make some prototypes.

- Consider always keeping available the last two point-releases; or the main (which is the last merged point-release) and the second to last point-release. This is meant to give enough freedom to users, in case one of the versions is not working as expected, in certain scenarios.

- Still on the topic of stability and trust to the public. Consider adding a second Developer branch (e.g., Developer-Green). This is not meant for developers to work on directly. Instead, it should always be the latest copy of the Developer branch triggered automatically when all Jenkins tests have passed, all are green. This would help teams/studios/developers to test the latest-developer-branch at its most-stable-state possible! It could also help track down (some) unknown issues that break the build during development that are not flagged during the tests; as well as allow others to test things before the point-release branches are available.

---

- Better UI borders of all editors, columns, buttons (life quality improvement). The colours are not the issue, it is the borders. If you notice Visual Studio for example, the borders, columns and buttons have more than 1 colour. The strain in the eyes are arguably less than in O3DE.
![UIwithBetterBorders](https://github.com/o3de/o3de/assets/44119535/7db7c851-5f36-4853-bd20-ad83647e40c0)
It gets even more noticeable when you stack multiple editors of O3DE on top of each other. Usually, having multi-colored borders help with this issue.

- Some tabs that were opened when O3DE crashes do not persist between sessions. Unsure if this is a security measure and intended (to prevent the same error from occurring when you reopen) or unintended. Recovery of the work-in-progress and state could be a good quality of life improvement. This reminded me of the early days of Visual Studio... when it crashed, all tabs and unsaved work went with it.

- Still in the O3DE editor, the inspector of components. Consider an option to collapse all components and re-open all components. Consider the persistence of these choices between sessions (collapsed or open). 
NOTE: I suspected someone started, but did not finish. The UI Editor and Lua Editor are re-opened in the next sessions, but none of the previous work/files are loaded again

--- 

- Editors as tabs instead of isolated forms (life quality improvement). Similar to Visual Studio, docking of the tabs/forms could improve productivity, especially for professionals working with one monitor only. Notice how the presenters struggle when trying to work with multiple editors:
https://www.youtube.com/watch?v=pUDPEUg7g9A&ab_channel=Open3DEngine&t=15m20s
I am not experienced with Qt (I think it is the tech used in the O3DE UI) but it should be something with this: https://doc.qt.io/qt-6/qtabwidget.html
The font used could also use some love, especially the Lua debugger (too much strain in my eyes)

- Still on docking of the panels. The user should have more freedom in how to dock and organize their panels. E.g., allow the middle to dock in the bottom without expanding the console from left to right... but instead, fill the available space on the bottom.
![image](https://github.com/o3de/o3de/assets/44119535/e85c4a0e-38b3-47e5-90a5-75d25a7b4792)

- Console. Provide more options such as word-wrap to prevent scrolling left and right on long string/text; allow a timestamp on the left; or allow lines-numbering.

--- 

- Publishing a game/prototype for Linux/Windows/Android. This is also something that should be addressed, documented and explained. Even if it is just a sketch or proof-of-concept. The objective is always to create a final product that can be used/installed by the users. E.g., the multiplayer sample could be available on Steam for free https://github.com/o3de/o3de-multiplayersample
I am assuming the multiplayer sample is capable of working as if one of the players is client-hosting the game, instead of a server-hosting.

DONE. SOLVED. Addressed here: https://www.youtube.com/watch?v=fV1gmxiRHdY&ab_channel=Open3DEngine

--- 

Some quality-of-life improvements similar to how Blender works.

- Option to change the camera view similar to Blender. Lock on X, Y or Z. There are occasions that it is beneficial to lock the camera on an axis. e.g., when adding colliders by hand on complex meshes.
Blender image:
![image](https://github.com/o3de/o3de/assets/44119535/9ab7d03d-5b25-4fef-be10-3d51f230b4d5)

- When you click on any entity's transform and start dragging it, allow the dragging to continue even if the mouse hits the end of the screen. E.g., when you have something selected and you reach the end of the screen on the right; you continue to move the mouse, and the cursor goes automatically to beggining of your screen on the left; continuously moving the cursor.

- Value inputs from components. Numbers and vectors should allow a calculation/formula to be inserted besides only-a-number. e.g., instead of just allowing the input 0.075, allow 0.15/2
![image](https://github.com/o3de/o3de/assets/44119535/236c5c9f-ec22-4341-9c63-64ad0c855b3c)

---

Unknowns and questions: 

- Does O3DE renderer always render all faces or does it only render the faces visible to the current camera? If it is rendering everything, this might be a good performance improvement.  e.g.: Unity 6, https://www.youtube.com/watch?v=8ZIdejTiXAE&ab_channel=Unity&t=17m50s

- Slices. Most slices were replaced/deprecated giving room to Prefabs. However, UI slices are still in use. What is the goal for the UI slices? Are these meant to stay or are they going to be replaced too?

- Prefabs. I did not find a way to extract the name and tags of a prefab while or after I spawned it. Forcing me to write additional code, e.g., I need to have three arrays: one for the prefab-references, other for the prefab-names, other for prefab-tags. Ideally, the developer should have a way to access the properties of the prefab after it was instanciated.
NOTE: I am writing this because I have failed to find a way to access the properties... so I am assuming there is no way to do it!

- Overall configurations and keybinds. I am unsure if this already exists, but I would prefer all of my configs and key binds to match other applications (e.g., O3DE and Blender camera movement is the same on both applications). Shifting between different software while maintaining the same keys can potentially increase productivity.
Changing the keybinds is done where? Consider adding the ability to change, save, and later load custom-made keybindings (e.g., create a keybinds.file that matches all shortcuts of Blender).

- o3de_manifest.json! Overall, all attributes should be explained but the ones that strike me as odd are the default_engines_folder and default_restricted_folder. What is the purpose of these two attributes?
In my tests, when I register more than one engine inside the same manifest, it will always try to use the first one it finds, instead of the one I try to execute.
So... Instead of using one manifest, I ended up using several copies of the manifest.json, one for each engine, and switching them whenever I wanted to switch engines... Leaving the engine's attributes obsolete, or I am doing something wrong?
The default_restricted_folder is also doing nothing? Even if you change its path (e.g., "C:/O3DE_23102/Restricted"), it will continue to add values to the old path: "C:/Users/userName/O3DE/Restricted/Projects/.."

- Visual Studio Community and Microsoft. Not knowledgeable enough to understand how O3DE and Microsoft contract works. However, their commitment to improving UE5 has been increasing. Couldn't O3DE also benefit from these commitments/work?
https://www.youtube.com/watch?v=c4gUOLbD7kU&ab_channel=MicrosoftVisualStudio
https://www.youtube.com/watch?v=GGz5BGPUgZw&ab_channel=MicrosoftVisualStudio

- Regarding Lua Editor. It works, and for as long as it works, keep it! But If it ever stops working, please discontinue it. Give priority to the existing extension for Visual Studio Code: https://marketplace.visualstudio.com/items?itemName=JeremyOng.o3de and priority to a Visual Studio Community extension: 
https://learn.microsoft.com/en-us/visualstudio/extensibility/starting-to-develop-visual-studio-extensions?view=vs-2022
I write most of my code/tests in Lua. The current Lua editor causes a strain in my eyes; it does not save the configurations/options between sessions; and lacks the expected options/features of an up-to-date IDE.
Personally, I would prefer to discontinue the O3DE Lua Editor, in favor of Visual Studio Community (because I also use it to build the engine and make occasional C++ test-gems). Or my second option, Visual Studio Code (since it runs on all OS).

---

Hope it helps

**Comment by mythrz** [comment]
overrides_demo.lua scripts are responsible for what? Could these be a factor for the faulty materials?

**Reply by Ulrick28** [reply]
Thanks for this thorough writeup! Agree with many of the points. For the Roadmap, the O3DE Foundation (O3DF) is working on a roadmap to be release sometime in January. We are aware of the docs issues and are looking for someone to volunteer to help us to close this gap.

**Comment by AMZN-alexpete** [comment]
The changes in the point release branch were just merge to development in this PR https://github.com/o3de/o3de/pull/17272

**Comment by mythrz** [comment]
##20240111 Point-release 2023102 https://github.com/o3de/o3de/commit/1204690397a825a12e4d2a122a9789e1b59cb3c8

---

- There are still some cases where arrays are not working as expected. DPE affected most if not all input arrays. 
https://github.com/o3de/o3de/issues/17208#issuecomment-1853997322
```
PrefabReferences =
{
	default = {SpawnableScriptAssetRef()},
	description = "Multiple PrefabReferences to spawn."
}
```

Took the liberty to open a new ticket/issue since the upper one was closed as solved, but arrays of SpawnableScriptAssetRef() are still not working. Tested with Lua: https://github.com/o3de/o3de/issues/17365#issue-2077190207

Trackers: https://github.com/o3de/o3de/pull/17441 https://github.com/o3de/o3de/issues/17404 https://github.com/o3de/o3de/pull/17457

---

- TextInput.uicanvas!! Encountered another odd issue regarding text inputs. This might be a niche for the time being, but more advanced UI and games will most likely encounter this in the future.
Both my custom-made uicanvas and the existing examples are not working during game-runtime. I suspect it has something to do with the lack of a feature that allows all-and-any keyboard inputs to be ON when the textinput is focused, and all other input-event-bindings turned OFF.
https://github.com/o3de/o3de/issues/17319#issue-2061534314

Use-case example. If your game is using keyboard to control your character (wasd) and you can open the map (m). That map is similar to google maps and has a textinput to write the city. ISSUE: I did not find any documentation or code explaining how Open 3D Engine shifts from game-runtime-only input-bindings to textinput.uicanvas UI-only input-bindings (writing text with keyboard)!

NOTE: If you test with the UI Editor, it will work on the UI Editor only. But it will not work in the game-runtime, making the example incomplete/unusable.
https://www.docs.o3de.org/docs/user-guide/interactivity/user-interface/components/interactive/components-textinput/

---

- Dynamic slices, somewhat related to the same issue: https://github.com/o3de/o3de/issues/17319#issue-2061534314
In this installation, opening my custom-made UI dynamic slices no longer gave an error. I suspect it was giving an error due to other dependencies, during the "DPE-bug-saga".

I will leave it here in case someone ever encounters the error but does not know how to reproduce it. Follow these steps:
Create a UI slice, convert it to dynamic. Repeat the process multiple times and use UiSpawner to have dynamic slices spawn other dynamic slices. If you try to use these slices in a different project AND engine, SOMETIMES, they do not work.
I suspect they save/use an instance ID when they are created; even if you re-process the slice it does not fix the issue. But, if you detach and re-create a dynamic slice it starts to work again.
I did not open an issue on this, because it is difficult to reproduce and slices have an unknown future ahead.
Are slices going to be deprecated too in favor of prefabs in the UI?

- Still regarding Slices. Not sure how to reproduce this error "[Error] (UI) - Deleting a UI element entity directly rather than using DestroyElement. Element is named 'Split'"! Every time I push to the dynamic slice, this error appears. I suspect I deleted something from the uicanvas using the delete button? If I ever manage to properly reproduce this, I will open a more detailed account.

---

O3DE editor is always listening to inputs even when it is not focused. 

- If you have two monitors. Leave the O3DE editor open on the first. Open a notebook/.text-file one the second. Write on the notepad and notice that pressing shift is recognized in both apps. It should only be recognized on the notepad.
![image](https://github.com/o3de/o3de/assets/44119535/0ce86e0f-a9a6-4d72-935c-5269223c7ff4)

I am not sure if someone ever fixed this issue because I cannot reproduce it anymore: https://github.com/o3de/o3de/issues/16387#issue-1806201194 or if it was just a coincidence during my workflow. But the two somehow allowed shortcuts to be active when they shouldn't be.

---

Pending Jobs should prevent or warn the user from running the engine when there are jobs that still need to be processed/finish.

- The first timers often try to run the engine right after they install and create their first project, unknowingly the downside of doing so.
- I also noticed that while scripting in Lua, if I modify a script; save it; and run the engine immediately; the engine does not wait for the pending changes to apply before running. Those who do not know or notice this behaviour might assume the code/logic is wrong, but in truth, the changes never reached the runtime-session.
